src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -594,98 +522,203 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
-        let fiber = cloneFiber(root.current, priority);
-        try {
-          while (fiber) {
-            // TODO: this is the only place where we recurse and it's unfortunate.
-            // (This may potentially get us into handleErrors() again.)
-            fiber = performUnitOfWork(fiber, true);
+          // If an error is thrown, propagate the error to the next boundary
+          const te = trapError(boundary, nextError);
+          if (te.boundary) {
+            // If a boundary is found, push trapped error onto array
+            nextTrappedErrors.push(te);
+          } else {
+            // Otherwise, we'll need to throw
+            firstUncaughtError = firstUncaughtError || te.error;
           }
-        } catch (nextError) {
-          // If it throws, propagate the error.
-          nextTrappedErrors = nextTrappedErrors || [];
-          nextTrappedErrors.push(trapError(boundary, nextError));
         }
-      });
+      }
+
+
+      // Now that we attempt to flush any work that was scheduled by the boundaries
+      // If this creates errors, they will be pushed to nextTrappedErrors and the loop will continue
+      try {
+        performTaskWorkUnsafe(true);
+      } catch (error) {
+        const failedUnitOfWork = nextUnitOfWork;
+        // Reset because it points to the error boundary:
+        nextUnitOfWork = null;
+        if (!failedUnitOfWork) {
+          // We shouldn't end up here because nextUnitOfWork
+          // should always be set while work is being performed.
+          throw error;
+        }
+        const trappedError = trapError(failedUnitOfWork, error);
+        if (nextTrappedErrors) {
+          nextTrappedErrors.push(trappedError);
+        } else {
+          nextTrappedErrors = [trappedError];
+        }
+      }
+
+      if (!nextTrappedErrors || firstUncaughtError) {
+        break;
+      }
     }
 
-    ReactCurrentOwner.current = null;
+    nextTrappedErrors = null;
+    knownBoundaries = new Set();
+    priorityContext = previousPriorityContext;
+    isHandlingErrors = false;
 
-    // Surface the first error uncaught by the boundaries to the user.
     if (firstUncaughtError) {
       // We need to make sure any future root can get scheduled despite these errors.
       // Currently after throwing, nothing gets scheduled because these fields are set.
       // FIXME: this is likely a wrong fix! It's still better than ignoring updates though.
       nextScheduledRoot = null;
       lastScheduledRoot = null;
-
-      // Throw any unhandled errors.
       throw firstUncaughtError;
     }
   }
 
+  function findClosestErrorBoundary(fiber : Fiber): Fiber | null {
+    let maybeErrorBoundary = fiber.return;
+    while (maybeErrorBoundary) {
+      if (maybeErrorBoundary.tag === ClassComponent) {
+        const instance = maybeErrorBoundary.stateNode;
+        if (typeof instance.unstable_handleError === 'function' &&
+            !knownBoundaries.has(maybeErrorBoundary)) {
```

Comment:
This line here is how I got it to work, and why I had to move this function into the scheduler.